proposals/nnnn-isolated-conformances.md

```diff
@@ -0,0 +1,571 @@
+
+```swift
+protocol GlobalLookup {
+  static func lookupByName(_ name: String) -> Self?
+}
+
+func hasNamed<T: GlobalLookup>(_: T.Type, name: String) async -> Bool {
+   return await Task.detached {
+     return T.lookupByName(name) != nil
+   }.value
+}
+```
+
+To prevent such problems in generic code, this proposal restricts the ability to send metatypes of type parameters (e.g., `T.Type`) across isolation domains. The above code, which is accepted in Swift 6 today, would be rejected by the proposed changes here with an error message like:
+
+```swift
+error: capture of non-sendable type 'T.Type' in closure
+```
+
+A function like `hasNamed` can express the need for a `Sendable` metatype by introducing a requirement `T: SendableMetatype` (to make the metatype `Sendable`), e.g.,
+
+```swift
+func hasNamed<T: GlobalLookup & SendableMetatype>(_: T.Type, name: String) async -> Bool {
+   return await Task.detached {
+     return T.lookupByName(name) != nil
+   }.value
+}
+```
+
+As with `Sendable`, an isolated conformance cannot be combined with a `SendableMetatype` constraint:
+
+```swift
+extension MyModelType: isolated GlobalLookup {
+  static func lookupByName(_ name: String) -> Self? { ... }
+}
+
+// error: isolated conformance of 'MyModelType' to 'MyModelType' cannot be used to 
+// satisfy conformance requirement for a `SendableMetatype` type parameter 'T'.
+if hasNamed(MyModelType.self, "root") { ... }
+```
+
+Protocol conformances can also be discovered dynamically with the `as?` and `is` operators. For example, one could try to produce an `any Equatable` from a value of unknown type in any isolation domain:
+
+```swift
+func tryEquatable(_ lhs: Any, rhs: Any) -> Bool {
+  if let eLHS = lhs as? any Equatable {
+	  // use Equatable.==
+  } else {
+    return false
+  }
+}
+```
+
+The `Any` value could contain `MyModelType`, in which case the conformance to `Equatable` will be isolated. In such cases, the `as?` operation will check whether the code is running on the executor associated with the conformance's isolation. If so, the cast can succeed; otherwise, the case will fail (and produce `nil`).
+
+## Detailed design
+
+The proposed solution describes the basic shape of isolated conformances and how they interact with the type system. This section goes into more detail on the data-race safety issues that arise from the introduction of isolated conformances into the language. Then it details three rules that, together, ensure freedom from data race safety issues in the presence of isolated conformances:
+
+1. An isolated conformance can only be used within its isolation domain.
+2. When an isolated conformance is used to satisfy a generic constraint `T: P`, the generic signature must not include either of the following constraints: `T: Sendable` or `T.Type: Sendable`. 
```

Comment:
I think this is leftover from an earlier iteration that used `T.Type: Sendable` rather than `T: SendableMetatype`

```suggestion
2. When an isolated conformance is used to satisfy a generic constraint `T: P`, the generic signature must not include either of the following constraints: `T: Sendable` or `T: SendableMetatype`. 
```

Reply:
Fixed, thank you!